app/client.jsx

```diff
@@ -14,11 +15,25 @@ const store = configureStore(initialState, browserHistory);
 const history = syncHistoryWithStore(browserHistory, store);
 const routes = createRoutes(store);
 
+/**
+ * Callback function handling frontend route changes.
+ */
+function onUpdate() {
+  // Prevent duplicate fetches when first loaded.
+  if (window.__INITIAL_STATE__ !== null) {
+    window.__INITIAL_STATE__ = null;
```

Comment:
In my opinion, `__INITIAL_STATE__` is used to create initial store and match the markups server rendered. So if you navigate along these paths:

`/`(server rendered): `onUpdate` is triggered first time. `__INITIAL_STATE__` contains all required data in this page. No need to refetch.
↓
`/dashboard`: `onUpdate` triggered second time. `__INITIAL_STATE__` doesn't contain required data in this page. Call `fetchComponentDataBeforeRender`.
↓
`/`: `onUpdate` triggered again.  `__INITIAL_STATE__` is outdated for this page. Call `fetchComponentDataBeforeRender`.

To summarize, after loading server rendered page,  `__INITIAL_STATE__` is useless. When navigating to other pages  `__INITIAL_STATE__` doesn't contain required data. When navigating back to the first page `__INITIAL_STATE__` is outdated and needs refetch.